libraries/AP_Compass/AP_Compass_MMC5xx3.cpp

```diff
@@ -0,0 +1,297 @@
+#define REG_CONTROL1        0x09
+#define REG_CONTROL2        0x0A
+
+// bits in REG_CONTROL0
+#define REG_CONTROL0_RESET  0x10
+#define REG_CONTROL0_SET    0x08
+#define REG_CONTROL0_TM     0x01
+
+// bits in REG_CONTROL1
+#define REG_CONTROL1_SW_RST 0x80
+#define REG_CONTROL1_BW0    0x01
+#define REG_CONTROL1_BW1    0x02
+
+#define MMC5883_ID 0x0C
+
+AP_Compass_Backend *AP_Compass_MMC5XX3::probe(AP_HAL::OwnPtr<AP_HAL::I2CDevice> dev,
+                                              bool force_external,
+                                              enum Rotation rotation)
+{
+    if (!dev) {
+        return nullptr;
+    }
+    AP_Compass_MMC5XX3 *sensor = new AP_Compass_MMC5XX3(std::move(dev), force_external, rotation);
+    if (!sensor || !sensor->init()) {
+        delete sensor;
+        return nullptr;
+    }
+
+    return sensor;
+}
+
+AP_Compass_MMC5XX3::AP_Compass_MMC5XX3(AP_HAL::OwnPtr<AP_HAL::Device> _dev,
+                                       bool _force_external,
+                                       enum Rotation _rotation)
+    : dev(std::move(_dev))
+    , force_external(_force_external)
+    , rotation(_rotation)
+{
+}
+
+bool AP_Compass_MMC5XX3::init()
+{
+    // get pointer to i2c bus semaphore
+    AP_HAL::Semaphore *sem = dev->get_semaphore();
+
+    // take i2c bus sempahore
+    WITH_SEMAPHORE(sem);
+
+    dev->set_retries(10);
+
+    uint8_t whoami;
+    if (!dev->read_registers(REG_PRODUCT_ID, &whoami, 1) ||
+        whoami != MMC5883_ID) {
+        // not a MMC5883
+        return false;
+    }
+
+    // reset sensor
+    dev->write_register(REG_CONTROL1, REG_CONTROL1_SW_RST);
+
+    // 5ms minimum startup time
+    hal.scheduler->delay(10);
+
+    if (!dev->write_register(REG_CONTROL1, REG_CONTROL1_BW0 | REG_CONTROL1_BW1)) {
+        return false;
+    } // 16 bit operation, 1.6ms measurement time
+
+    /* register the compass instance in the frontend */
+    dev->set_device_type(DEVTYPE_MMC5883);
+    if (!register_compass(dev->get_bus_id(), compass_instance)) {
+        return false;
+    }
+
+    set_dev_id(compass_instance, dev->get_bus_id());
+
+    printf("Found a MMC5883 on 0x%x as compass %u\n", dev->get_bus_id(), compass_instance);
+
+    set_rotation(compass_instance, rotation);
+
+    if (force_external) {
+        set_external(compass_instance, true);
+    }
+
+    dev->set_retries(1);
+
+    // call timer() at 1kHz
+    dev->register_periodic_callback(1000,
+                                    FUNCTOR_BIND_MEMBER(&AP_Compass_MMC5XX3::timer, void));
+
+    return true;
+}
+
+void AP_Compass_MMC5XX3::timer()
+{
+    // recalculate the offset with set/reset operation every measure_count_limit measurements
+    // sensor is read at about 500Hz, so about every 10 seconds
+    const uint16_t measure_count_limit = 5000;
+    const uint16_t zero_offset = 32768; // 16 bit mode
+    const uint16_t sensitivity = 4096; // counts per Gauss, 16 bit mode
+    const float counts_to_milliGauss = 1.0e3f / sensitivity;
```

Comment:
`constexpr`?  Compiler will probably work it out.

Reply:
done